torch/utils/_pytree.py

```diff
@@ -274,7 +274,19 @@ def tree_flatten(pytree: PyTree) -> Tuple[List[Any], TreeSpec]:
 
 def tree_leaves(pytree: PyTree) -> List[Any]:
     """Get a list of leaves of a pytree."""
-    return tree_flatten(pytree)[0]
+    if _is_leaf(pytree):
+        return [pytree]
+
+    node_type = _get_node_type(pytree)
+    flatten_fn = SUPPORTED_NODES[node_type].flatten_fn
+    child_pytrees, context = flatten_fn(pytree)
+
+    # Recursively flatten the children
+    result : List[Any] = []
+    for child in child_pytrees:
```

Comment:
nit: maybe it's better to have an internal `_tree_flatten` function that optionally computes tree spec. Then wrap it in the public APIs.

Likewise for `tree_structure`

Reply:
I tried this and every way I tried made `tree_leaves` slower and none made `tree_structure` significantly faster so I think this is a reasonable position.